### ISSUE TITLE
High CPU usage

### ISSUE THREAD

**Issue by meonlineonmess** [kotcrab/vis-editor#240]
Hello,

Firstly, I would like to congratulate the creator for the VisEditor application which is helping the development of games in libGDX a lot !

I would like to report a bug which seems to affect the CPU usage which goes very high ~100% when the editor is not in focus or minimized. I use macOS Sierra and I am having CPU usage issues when switching to another application. The only way I can limit the CPU usage is using the app called AppPolice with which I can limit the CPU usage to 20%.

Overlap2D has the same issue like VisEditor.

Could it be related to the FPS ? this is what they said on Overlap2D forums.

I am looking forward for a fix to this issue.

Thank you very much.
Yours sincerely,
Ben

**Comment by bruceloco** [comment]
@meonlineonmess Lwjgl3 is the library used here, the main does not set FPS at all and VSync is set to true, also nothing for iddle time is set, so try
config.useVsync(false);
config.setIdleFPS(2); <- possibly don't do anything for .5 seconds

Don't use 0 for idle FPS as the code is unprotected:
Thread.sleep(1000 / config.idleFPS); <- deep inside the library

**Comment by bruceloco** [comment]
also there is no frames per second limit in lwjgl3
there is a thread here on how to limit it, but it would need to be implemented in the actual library
http://forum.lwjgl.org/index.php?topic=5653.0

**Comment by bruceloco** [comment]
@meonlineonmess did you try the changes?

**Comment by kotcrab** [comment]
I updated gdx version and added `setIdleFps` to try help this. Would be nice if anyone can try and tell me if it works or not. Javadocs tells that this sets fps for non-continous rendering mode so this may not change anything at all.

This isn't VisEditor issue, I've seen similar behaviour in other applications, also on Windows 7 and 10. In LWJGL you could set `backgroundFps` which fixed this problem but same options seems to be missing in LWJGL3.

**Comment by bruceloco** [comment]
@kotcrab I opened a bug: https://github.com/libgdx/libgdx/issues/4526
This is not the only problem. 
If you don't limit the foreground speed, you could get 250fps per second or depends on how much can lwjgl3 can do on your machine. 
Not much of a problem on a desktop, but on a laptop it is overkill, overhog and battery draining.
I checked the lwjgl library(keep in mind that I only know about this things for 1.5 weeks) and I did not see a foregroundFPS setting as existed in lwjgl
I'll pull from your repo and build on my mac to see

**Comment by kotcrab** [comment]
VisEditor has vsync enabled so it doesn't render more than 60 FPS. You can actually try to disable it and you will see that it will then render as many frames as possible.

**Comment by bruceloco** [comment]
Some guys were complaining about vsync in OSX.
Personally I did not see anything and I see 60 fps at the bottom of vis-edit

**Comment by bruceloco** [comment]
Ok, I see it now

Foreground app:
CPU 6-10%
GPU 60fps - similar load on the gpu processor

Background app(minimised):
CPU 90-95%
GPU 1-2fps - full load

Vsync on/off does nothing to address the issue, it will only increase fps(in my system to about 250, as @kotcrab mentioned.
setIdleFPS also does not have an effect on the CPU and GPU usage, only on the fps when in background.

OSX sierra
Gradle 3.3
Javavm 1.8

**Comment by bruceloco** [comment]
Hi @kotcrab 
I think I found the issue, it's in libgdx:

In backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/Lwjgl3Application.java 
they are using:
		GLFW.glfwPollEvents();
instead of 
		GLFW.glfwWaitEvents();

Check this thread:
https://github.com/libgdx/libgdx/issues/4419

Any chance you can build pull a branch to build libgdx to just test that backend change?

**Comment by bruceloco** [comment]
just tested the change
so after this the framecount drops but so does all the processing, so fps is irregular.
the good news is that when minimised, cpu and gpu usage disappears, about 1%

**Comment by bruceloco** [comment]
ok, after some investigation and tweaking:
I got it to run at 60fps when normal and almost nothing when minimised.
required a small modification in:
public class Lwjgl3Window implements Disposable {
switch 	private boolean iconified = false; to public
in the main loop in lwjgl3application:
			boolean haveWindowsRendered = false;
			closedWindows.clear();
			boolean is_minimised=false;
			boolean window_zero = true;
			for (Lwjgl3Window window : windows) {
				window.makeCurrent();
				currentWindow = window;
				if(window_zero && window.iconified) is_minimised = true;

				window_zero = false;
				synchronized (lifecycleListeners) {
					haveWindowsRendered |= window.update();
				}
				if (window.shouldClose()) {
					closedWindows.add(window);
				}
			}

			if (is_minimised)
			{
				GLFW.glfwWaitEvents();
			} else {
				GLFW.glfwPollEvents();
			}

**Comment by WickedShell** [comment]
I imagine that that change set would break non-continous rendering as
rendering is expected to resume when a runnable is posted rather then
waiting for GLFW event.

**Comment by bruceloco** [comment]
Unsure, but the window is minimised, so no need to render. For the purposes of an editor it is fine.
The impact on a game I can't tell(don't have any).
However when a game is minimised isn't it something that should be paused? Hence not rendering?
This was my first 3 hour attempt at solving an issue in code I don't know. 
Hoping that someone with experience can take a swing at it ;-)

**Comment by bruceloco** [comment]
Quoting the user from another thread:
"You can wake it up from glfwWaitEvents() with glfwPostEmptyEvent()."

**Comment by bruceloco** [comment]
i'm glad to share my lwjgl3 backend jar, it's easy to replace on the gradle cach
These on OSX are in home/.gradle/
Do a find to locate the relevant file and replace it

**Comment by bruceloco** [comment]
[gdx-backend-lwjgl3-1.9.5.jar.zip](https://github.com/kotcrab/vis-editor/files/705939/gdx-backend-lwjgl3-1.9.5.jar.zip)

here's the build lib
from mac:
find .gradle/ -name gdx-backend-lwjgl3* 
to locate where gradle put the library it downloaded.
Then replace it with the attached one

**Comment by bruceloco** [comment]
here's a snapshot for people to test:
https://www.dropbox.com/s/2myvskp6hd9y4n9/VisEditor-0.4.0-SNAPSHOT.tar?dl=0

**Comment by bruceloco** [comment]
@kotcrab did you manage to test this on windows?

**Comment by bruceloco** [comment]
tested on windows, did not see any side effects

**Comment by WimpyMistake** [comment]
@bruceloco i tried your build on my mac, but the CPU is still high when editor is lost focus. Im on macOS sierra, 10.12.2, 2.4 GHz Intel Core i7

The app police trick does work though. thanks @meonlineonmess

**Comment by bruceloco** [comment]
@WimpyMistake I did not do any lib changes for out of focus, only minimised.
Explain this out of focus CPU usage, is it high like 100% high? also are we talking GPU? CPU?

**Comment by WimpyMistake** [comment]
@bruceloco sorry i didn't check the gpu, only cpu. it was 90% plus but the fan noise is very loud. its the same behaviour with original VisEditor.

**Comment by bruceloco** [comment]
@WimpyMistake 
Does it happen if you do:
 java -XstartOnFirstThread -jar vis-editor-0.4.0-SNAPSHOT.jar

**Comment by WimpyMistake** [comment]
@bruceloco yes, its still happening. but i confirmed if minimised it works well.
p/s how do you check for gpu usage? i use activity monitor to check CPU but no GPU option.

**Comment by bruceloco** [comment]
@WimpyMistake fixing that one is complicated.
So when minimised there is no need to render on screen, but when out of focus, the rendering should still continue.
I haven't checked yet where. Also remember that I am modifying a part of libgdx, so far I haven't had any traction on the issue that I opened.
There is the assumption here that for the editor we might not care so I can use the same fix that I did for the minimised.
I believe there is a bug that when an app is out of focus it does not check for vsync and attempts to render at full speed.
I need to review my fix and look for the vsync aspect of libdgx
